backbone/src/main/java/org/researchstack/backbone/ui/step/layout/ConsentReviewSubstepListStepLayout.java

```diff
@@ -0,0 +1,128 @@
+package org.researchstack.backbone.ui.step.layout;
+
+import android.content.Context;
+import android.util.AttributeSet;
+
+import org.researchstack.backbone.DataProvider;
+import org.researchstack.backbone.model.ConsentSignatureBody;
+import org.researchstack.backbone.model.ProfileInfoOption;
+import org.researchstack.backbone.model.User;
+import org.researchstack.backbone.model.survey.factory.SurveyFactory;
+import org.researchstack.backbone.result.StepResult;
+import org.researchstack.backbone.result.TaskResult;
+import org.researchstack.backbone.utils.ObservableUtils;
+import org.researchstack.backbone.utils.StepResultHelper;
+
+import java.util.Date;
+
+/**
+ * Created by TheMDP on 1/16/17.
+ *
+ * Consent ReviewStep contains a number of steps
+ */
+
+public class ConsentReviewSubstepListStepLayout extends ViewPagerSubstepListStepLayout {
+
+    public ConsentReviewSubstepListStepLayout(Context context) {
+        super(context);
+    }
+
+    public ConsentReviewSubstepListStepLayout(Context context, AttributeSet attrs) {
+        super(context, attrs);
+    }
+
+    @Override
+    protected void onComplete() {
+        ConsentSignatureBody consentSignatureBody = createConsentSignatureBody(stepResult, taskResult);
+        if (DataProvider.getInstance().isSignedIn(getContext())) {
+            uploadConsent(consentSignatureBody);
+        } else {
+            DataProvider.getInstance().saveConsent(getContext(), consentSignatureBody);
+            super.onComplete();
+        }
+    }
+
+    protected void uploadConsent(ConsentSignatureBody consentSignatureBody) {
+        DataProvider.getInstance()
+                .uploadConsent(getContext(), consentSignatureBody)
+                .compose(ObservableUtils.applyDefault())
+                .subscribe(dataResponse -> {
```

Comment:
@liujoshua I ended up fixing this, see StepLayoutHelper method safePerform, and how it's used in LoginStepLayout.